File: ios/LLMHub/Sources/LLMHub/MimoBot/Kokoro/KokoroAssets.swift
import Foundation

/// Manages on-device Kokoro asset files. Downloads from the canonical
/// onnx-community Hugging Face mirror on first use and caches under
/// `Documents/kokoro/`.
///
/// Files (one-time download):
///   - model_fp16.onnx           (≈ 165 MB)
///   - voices/<voice>.bin        (≈ 512 KB each)
enum KokoroAssets {

    struct Progress: Sendable {
        let stage: String
        let bytesDone: Int64
        let bytesTotal: Int64   // -1 if unknown
    }

    private static let modelURL = URL(string:
        "https://huggingface.co/onnx-community/Kokoro-82M-v1.0-ONNX/resolve/main/onnx/model_fp16.onnx"
    )!

    private static func voiceURL(_ id: String) -> URL {
        URL(string:
            "https://huggingface.co/onnx-community/Kokoro-82M-v1.0-ONNX/resolve/main/voices/\(id).bin"
        )!
    }

    static func root() throws -> URL {
        let docs = try FileManager.default.url(
            for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true
        )
        let dir = docs.appendingPathComponent("kokoro", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    static func modelFile() throws -> URL {
        try root().appendingPathComponent("model_fp16.onnx")
    }

    static func voiceFile(_ id: String) throws -> URL {
        let voicesDir = try root().appendingPathComponent("voices", isDirectory: true)
        try FileManager.default.createDirectory(at: voicesDir, withIntermediateDirectories: true)
        return voicesDir.appendingPathComponent("\(id).bin")
    }

    static func isReady(voice id: String) -> Bool {
        guard let model = try? modelFile(), let voice = try? voiceFile(id) else { return false }
        return FileManager.default.fileExists(atPath: model.path) &&
               FileManager.default.fileExists(atPath: voice.path)
    }

    /// Download model + voice if missing. Yields progress + a final "Ready" event.
    static func ensure(voice id: String = "af_heart") -> AsyncThrowingStream<Progress, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let model = try modelFile()
                    if !FileManager.default.fileExists(atPath: model.path) {
                        continuation.yield(Progress(stage: "Downloading Kokoro model", bytesDone: 0, bytesTotal: -1))
                        try await download(from: modelURL, to: model) { done, total in
                            continuation.yield(Progress(stage: "Downloading Kokoro model", bytesDone: done, bytesTotal: total))
                        }
                    }
                    let voice = try voiceFile(id)
                    if !FileManager.default.fileExists(atPath: voice.path) {
                        continuation.yield(Progress(stage: "Downloading voice \(id)", bytesDone: 0, bytesTotal: -1))
                        try await download(from: voiceURL(id), to: voice) { done, total in
                            continuation.yield(Progress(stage: "Downloading voice \(id)", bytesDone: done, bytesTotal: total))
                        }
                    }
                    continuation.yield(Progress(stage: "Ready", bytesDone: 1, bytesTotal: 1))
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    private static func download(
        from url: URL,
        to target: URL,
        onProgress: @escaping (_ done: Int64, _ total: Int64) -> Void
    ) async throws {
        var request = URLRequest(url: url)
        request.timeoutInterval = 60 * 30  // 30 min for big model
        let (asyncBytes, response) = try await URLSession.shared.bytes(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw NSError(domain: "KokoroAssets", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "HTTP error for \(url)"])
        }
        let total = response.expectedContentLength
        let tmp = target.appendingPathExtension("part")
        FileManager.default.createFile(atPath: tmp.path, contents: nil)
        let handle = try FileHandle(forWritingTo: tmp)
        defer { try? handle.close() }

        var done: Int64 = 0
        var lastEmit: Int64 = 0
        var buf = Data()
        buf.reserveCapacity(64 * 1024)

        for try await byte in asyncBytes {
            buf.append(byte)
            if buf.count >= 64 * 1024 {
                try handle.write(contentsOf: buf)
                done += Int64(buf.count)
                buf.removeAll(keepingCapacity: true)
                if done - lastEmit >= 1_000_000 {
                    onProgress(done, total)
                    lastEmit = done
                }
            }
        }
        if !buf.isEmpty {
            try handle.write(contentsOf: buf)
            done += Int64(buf.count)
        }
        try handle.close()

        // Atomically rename to the final path.
        if FileManager.default.fileExists(atPath: target.path) {
            try FileManager.default.removeItem(at: target)
        }
        try FileManager.default.moveItem(at: tmp, to: target)
        onProgress(done, total)
    }
}

File: android/app/src/main/java/com/llmhub/llmhub/mimobot/speech/kokoro/KokoroAssets.kt
package com.llmhub.llmhub.mimobot.speech.kokoro

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Manages on-device Kokoro asset files. Downloads from the canonical
 * onnx-community Hugging Face mirror on first use and caches under
 * `filesDir/kokoro/`.
 *
 * Files (one-time download):
 *   - kokoro-v1.0.fp16.onnx  (≈ 165 MB)
 *   - voices/<voice>.bin     (≈ 512 KB each)
 *
 * Reuses OkHttp which is already a project dependency.
 */
object KokoroAssets {

    private const val TAG = "KokoroAssets"

    // The fp16 export is small enough for phones, fast enough on NNAPI EP.
    private const val MODEL_URL =
        "https://huggingface.co/onnx-community/Kokoro-82M-v1.0-ONNX/resolve/main/onnx/model_fp16.onnx"

    private fun voiceUrl(voiceId: String) =
        "https://huggingface.co/onnx-community/Kokoro-82M-v1.0-ONNX/resolve/main/voices/$voiceId.bin"

    private val http: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.MINUTES)   // big model download
            .writeTimeout(2, TimeUnit.MINUTES)
            .build()
    }

    fun root(context: Context): File =
        File(context.filesDir, "kokoro").also { it.mkdirs() }

    fun modelFile(context: Context): File = File(root(context), "model_fp16.onnx")
    fun voiceFile(context: Context, voiceId: String): File =
        File(File(root(context), "voices").also { it.mkdirs() }, "$voiceId.bin")

    fun isReady(context: Context, voiceId: String): Boolean =
        modelFile(context).exists() && voiceFile(context, voiceId).exists()

    /** Progress event emitted during download. [bytesDone] / [bytesTotal] in bytes; -1 unknown. */
    data class Progress(
        val stage: String,
        val bytesDone: Long,
        val bytesTotal: Long,
    )

    /**
     * Download model + voice if missing. Emits one [Progress] per ~1 MB of data
     * plus stage transitions. Caller can collect this Flow to drive a UI.
     */
    fun ensure(context: Context, voiceId: String = "af_heart"): Flow<Progress> = flow {
        val mf = modelFile(context)
        val vf = voiceFile(context, voiceId)

        if (!mf.exists()) {
            emit(Progress("Downloading Kokoro model", 0, -1))
            download(MODEL_URL, mf) { done, total ->
                emit(Progress("Downloading Kokoro model", done, total))
            }
        }
        if (!vf.exists()) {
            emit(Progress("Downloading voice $voiceId", 0, -1))
            download(voiceUrl(voiceId), vf) { done, total ->
                emit(Progress("Downloading voice $voiceId", done, total))
            }
        }
        emit(Progress("Ready", 1, 1))
    }.flowOn(Dispatchers.IO)

    private suspend fun download(
        url: String,
        target: File,
        onProgress: suspend (done: Long, total: Long) -> Unit,
    ) {
        val req = Request.Builder().url(url).build()
        val tmp = File(target.parentFile, target.name + ".part")
        try {
            http.newCall(req).execute().use { resp ->
                if (!resp.isSuccessful) throw IllegalStateException("HTTP ${resp.code} for $url")
                val body = resp.body ?: throw IllegalStateException("empty body for $url")
                val total = body.contentLength()
                var done = 0L
                var lastEmit = 0L
                tmp.outputStream().use { out ->
                    val src = body.source()
                    val buf = ByteArray(64 * 1024)
                    while (true) {
                        val n = src.read(buf)
                        if (n <= 0) break
                        out.write(buf, 0, n)
                        done += n
                        if (done - lastEmit >= 1_000_000L) {
                            onProgress(done, total)
                            lastEmit = done
                        }
                    }
                }
            }
            if (!tmp.renameTo(target)) {
                tmp.copyTo(target, overwrite = true)
                tmp.delete()
            }
            Log.i(TAG, "downloaded ${target.absolutePath} (${target.length()} bytes)")
        } catch (t: Throwable) {
            tmp.delete()
            throw t
        }
    }
}
